src/gateway/GatewayV2.sol

```diff
@@ -0,0 +1,343 @@
+            require(activeRouters[router].id != 0, "Gateway/invalid-router");
+            recoveries[messageHash] = Recovery(block.timestamp + RECOVERY_CHALLENGE_PERIOD, router);
+            emit InitiateMessageRecovery(messageHash, router);
+        } else if (MessagesLib.messageType(payload) == MessagesLib.Call.DisputeMessageRecovery) {
+            bytes32 messageHash = payload.toBytes32(1);
+            return _disputeMessageRecovery(messageHash);
+        }
+    }
+
+    /// @inheritdoc IGatewayV2
+    function disputeMessageRecovery(bytes32 messageHash) external auth {
+        _disputeMessageRecovery(messageHash);
+    }
+
+    function _disputeMessageRecovery(bytes32 messageHash) internal {
+        delete recoveries[messageHash];
+        emit DisputeMessageRecovery(messageHash);
+    }
+
+    /// @inheritdoc IGatewayV2
+    function executeMessageRecovery(bytes calldata message) external {
+        bytes32 messageHash = keccak256(message);
+        // wouldn't it better to mark these as memory?
+        Recovery storage recovery = recoveries[messageHash];
+        Router storage router = activeRouters[recovery.router];
+
+        require(recovery.timestamp != 0, "Gateway/message-recovery-not-initiated");
+        require(recovery.timestamp <= block.timestamp, "Gateway/challenge-period-has-not-ended");
+        require(router.id != 0, "Gateway/invalid-router");
+
+        delete recoveries[messageHash];
+        _handle(message, recovery.router, router, true);
+        emit ExecuteMessageRecovery(message);
+    }
+
+    // --- Outgoing ---
+
+    /// @inheritdoc IGatewayV2
+    function send(bytes calldata message, address source) public payable auth {
+        bytes memory proof = abi.encodePacked(uint8(MessagesLib.Call.MessageProof), keccak256(message));
+
+        uint256 numRouters = routers.length;
+        require(numRouters > 0, "Gateway/routers-not-initialized");
+        uint256 fuel = allowance;
+        uint256 tank = fuel > 0 ? fuel : gasService.shouldRefuel(source, message) ? address(this).balance : 0;
+
+        uint256 destChainMsgCost = gasService.estimate(message);
+        uint256 destChainProofCost = gasService.estimate(proof);
+
+        for (uint256 i; i < numRouters; i++) {
+            RouterLike currentRouter = RouterLike(routers[i]);
+            bool isPrimaryRouter = i == PRIMARY_ROUTER_ID - 1;
+            bytes memory payload = isPrimaryRouter ? message : proof;
+
+            uint256 consumed = currentRouter.estimate(payload, isPrimaryRouter ? destChainMsgCost : destChainProofCost);
+            uint256 remaining;
+            unchecked {
+                remaining = tank - consumed;
+            }
+            require(remaining < tank, "Gateway/not-enough-gas-funds");
+            tank = remaining;
+
+            currentRouter.pay{value: consumed}(payload, address(this));
+            currentRouter.send(payload);
+        }
+
+        if (fuel > 0 && tank > 0) allowance = 0;
```

Comment:
Looks to me like the separate variable is not really needed here.
```suggestion
        uint256 tank = allowance > 0 ? allowance : gasService.shouldRefuel(source, message) ? address(this).balance : 0;

        uint256 destChainMsgCost = gasService.estimate(message);
        uint256 destChainProofCost = gasService.estimate(proof);

        for (uint256 i; i < numRouters; i++) {
            RouterLike currentRouter = RouterLike(routers[i]);
            bool isPrimaryRouter = i == PRIMARY_ROUTER_ID - 1;
            bytes memory payload = isPrimaryRouter ? message : proof;

            uint256 consumed = currentRouter.estimate(payload, isPrimaryRouter ? destChainMsgCost : destChainProofCost);
            uint256 remaining;
            unchecked {
                remaining = tank - consumed;
            }
            require(remaining < tank, "Gateway/not-enough-gas-funds");
            tank = remaining;

            currentRouter.pay{value: consumed}(payload, address(this));
            currentRouter.send(payload);
        }

        if (allowance > 0 && tank > 0) allowance = 0;
```

Reply:
I think `destChainMsgCost` and `destChainProofCost` here can just be `messageCost` and `proofCost`

Reply:
Variables  changed.
`fued` variable is used to cache the storage one in order to save gas from reads.